v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,240 @@
+package logs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/open-policy-agent/opa/v1/plugins/rest"
+)
+
+// eventBuffer stores and uploads a gzip compressed JSON array of EventV1 entries
+type eventBuffer struct {
+	buffer               chan []byte    // buffer stores JSON encoded EventV1 data
+	mtx                  sync.RWMutex   // mtx controls access to the buffer
+	payload              *payloadBuffer // payload contains the compressed JSON array to be uploaded
+	client               rest.Client    // client is used to upload the data to the configured service
+	uploadPath           string         // uploadPath is the configured HTTP resource path for upload
+	uploadSizeLimitBytes int64          // uploadSizeLimitBytes will enforce a maximum payload size to be uploaded
+}
+
+func newEventBuffer(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) *eventBuffer {
+	return &eventBuffer{
+		buffer:               make(chan []byte, bufferSizeLimitEvents),
+		payload:              newPayloadBuffer(),
+		client:               client,
+		uploadPath:           uploadPath,
+		uploadSizeLimitBytes: uploadSizeLimitBytes,
+	}
+}
+
+// Reconfigure updates the user configurable values
+func (b *eventBuffer) Reconfigure(bufferSizeLimitEvents int64, client rest.Client, uploadPath string, uploadSizeLimitBytes int64) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	b.client = client
+	b.uploadPath = uploadPath
+	b.uploadSizeLimitBytes = uploadSizeLimitBytes
+
+	if int64(cap(b.buffer)) == bufferSizeLimitEvents {
+		return
+	}
+
+	close(b.buffer)
+	newBuffer := make(chan []byte, bufferSizeLimitEvents)
+
+	for event := range b.buffer {
+		if int64(len(event)) < b.uploadSizeLimitBytes {
+			push(newBuffer, event)
+			continue
+		}
+
+		var decoded EventV1
+		if err := json.Unmarshal(event, &decoded); err != nil {
+			continue
+		}
+		if err := b.dropNDCache(&decoded, event); err != nil {
+			if !errors.As(err, &droppedNDCache{}) {
+				continue
+			}
+			event, err = json.Marshal(decoded)
+			if err != nil {
+				continue
```

Comment:
[question] Is this so unlikely (marshalling errors on reconfigure) that we can silently drop the event? I'd expect an error log of some sort, as right now, the user will have no real chance to figure out that something unexpected happened. I suppose it _is_ very unlikely, though 🤔 But I think it still wouldn't hurt to error-log something? What do you think?

Reply:
I agree I think logging this error is worth doing. This made me realize that the `droppedNDCache` error also needs to be handled so that the metrics can be incremented. Updated it so that the errors are recorded so they can be logged.

While at the moment I don't think a Marshal error is possible when re-configuring because a invalid event couldn't get into the channel, that might change in the future so good to handle it now.